### DISCUSSION TITLE
How to create and use custom types?

### DISCUSSION THREAD

**Discussion by niklasfjeldberg** [nuxt/nuxt#19080] [Questions]
How do you create and use custom types to use with Typescript in Nuxt 3? If you have the file `/types/custom-types.ts` with the following content:
```
export interface Recipe {
  authors: any[],
  commentsCount: number,
  ...
}
```

Then the following in any `.vue` file:
```
import type { Recipe } from '~~/types/custom-types';
findOne<Recipe>(dataType, dataID)
```

Should this not work?

**Comment by AloisSeckar** [comment]
I see no problem with that code (except ofc I don't know what `findOne` should do). But the type import itself works for me. You are running into some troubles? Can you provide more details?

**Reply by niklasfjeldberg** [reply]
@AloisSeckar thank you for your prompt reply. I shall look into my code some more then. I will provide more information if I cannot solve it. `findOne` is a function of the package `@nuxtjs/strapi`.

I am using it the following way, and I get this error `Property 'x' does not exist on type 'Strapi4ResponseData<Recipe>'.ts(2339)` in my code when accessing `response.data.x` in the `<template>.`
```
const { data: response } = await useAsyncData(fetchName, () => findOne<Recipe>(dataType, dataID, {
  populate: [
    'image',
    ...
  ]
}));
```

**Reply by AloisSeckar** [reply]
Still can't tell for sure without seeing the full picture. But according to my understanding of [the repo](https://github.com/nuxt-modules/strapi/blob/dev/src/runtime/types/v4.d.ts), Strapi4ResponseData should contain the actuall data-object inside `attributes` property.

When I work with something new, I first try to console.log() (or you can peek through Vue.js devtools, if you're not a cave man like me :D) the entire object to see, how it is structured inside. It often helps a lot.

**Reply by niklasfjeldberg** [reply]
I believe I found out what's wrong. I have the plugin [Transformer](https://market.strapi.io/plugins/strapi-plugin-transformer) on my Strapi backend, and that removes attributes (among other stuff) from the responses. So, when I am using `findOne<Recipe>` I am using the predefined type `Strapi4ResponseSingle` which takes my input of `<Recipe>`. Which comes down to:

```
export interface Strapi4ResponseData<T> {
  id: number,
  attributes: T,
  meta: Record<string, unknown>
}
```

So, my code is looking for the attribute's property, which does not exist. Therefore my code is looking for my properties in the wrong place, making it so that they "do not exist". I am a bit unsure how to redefine the type for `findOne`, but I believe that is what I need to do.

@AloisSeckar thanks for the input!
